### DISCUSSION TITLE
How do I see or what gets invoked if someone changes values in the server?

### DISCUSSION THREAD

**Discussion by jnsw** [eclipse-milo/milo#930] [General]
I am using the server implementation and want to do stuff, if someone changes values on the server. Do we have a callback routine or something for this?

**Comment by kevinherron** [comment]
Your AddressSpace will receive a call to the `write` method.

If you’re using a ManagedAddressSpace like the example you can install an `AttributeFilter` or `AttributeObserver` on the Node you’re interested in. Or override the `write` method and make sure to call `super` when done looking at what is happening.

**Reply by jnsw** [reply]
Thanks, that worked like a charm. 
I have another question: Can we know if the fired AttributeChange is from the "outside", so an OPC UA Client changed the value instead of ourselfes?

**Reply by kevinherron** [reply]
If you’re using the AttributeFilter mechanism the context will have an Optional containing a Session on it.

**Comment by MiproHTy** [comment]
I've been banging my head to get this to work. Can someone please show example how the AttributeFilter is going to need to be implemented and how values are updated to server from the software itself so that it doesnt trigger the AttributeChange?

The only time i got this working is when using
`misoNodes.get(i).getFilterChain().addLast(
				            new AttributeLoggingFilter(),
				            AttributeFilters.getValue(
				            		ctx ->
				                    	new DataValue(new Variant(value))
				            )
				        );`
Thats when AttributeChange doesnt trigger BUT when i'm updating about 100 variables in server in a loop which sleeps every 500ms i get StackOverFlow in about 10 seconds. So i think im using this wrong and there is correct way to do this?

**Reply by kevinherron** [reply]
You don't need an `AttributeFilter` if you are updating the values by some other means (your loop in this case). `AttributeFilter` is a way to "intercept" calls to get the value of an attribute, either for a Node that you don't fully "control" or didn't instantiate, or that you aren't setting the attribute values on by some other means.

**Reply by MiproHTy** [reply]
Thank you for still answering :)

My problem was that when using AttributeObserver on nodes what i created i get AttributeChanged even when i write myself the value to node. My ugly workaround is now this:
`Nodes.get(i).removeAttributeObserver(myObs);`
`Nodes.get(i).setValue(new DataValue(new Variant(value)));`
`Nodes.get(i).addAttributeObserver(myObs);`

So when i update myself the values in the server my observer will not fire and if client writes i get the attributechanged.
Do you have suggestions how this should have been done?

**Reply by kevinherron** [reply]
No, it's working as intended. Why does your `AttributeObserver` being notified of the attribute change lead to an infinite loop or stack overflow?

**Reply by MiproHTy** [reply]
`Nodes.get(i).getFilterChain().addLast( new AttributeLoggingFilter(), AttributeFilters.getValue( ctx -> new DataValue(new Variant(value)) ) );`
If you run this code in a loop it stackoverflows.

**Reply by kevinherron** [reply]
Why would you continue to add filters in a loop? That’s also a memory leak.
